File: sdk/go/env_test.go
package shojiku

import (
	"context"
	"os"
	"path/filepath"
	"slices"
	"strings"
	"testing"
)

func TestLookupsOffStripEverySHOJIKUVariableFromTheChild(t *testing.T) {
	// The clause a subprocess SDK owes that the in-process ones do not: the
	// engine reads SHOJIKU_FONT_DIR and SHOJIKU_LOCALE_DIR itself, so a
	// client that merely stopped reading them would have moved the lookup one
	// process away rather than closed it.
	e := &env{enabled: false, source: []string{
		"SHOJIKU_FONT_DIR=/fonts",
		"SHOJIKU_LOCALE_DIR=/locales",
		"SHOJIKU_TEMPLATE_ROOT=/templates",
		"PATH=/usr/bin",
		"HOME=/root",
	}}

	child := e.childEnvironment()

	for _, entry := range child {
		if strings.HasPrefix(entry, envPrefix) {
			t.Errorf("the child inherited %q", entry)
		}
	}
	if !slices.Contains(child, "HOME=/root") {
		t.Error("an unrelated variable was stripped too")
	}
}

func TestLookupsOffDoNotStripPath(t *testing.T) {
	// PATH is the operating system's variable, not one of this engine's
	// settings. Gating it would leave a hermetic client unable to run an
	// installed binary at all.
	e := &env{enabled: false, source: []string{"PATH=/usr/bin", "SHOJIKU_BIN=/nope"}}

	if got := e.unguarded("PATH"); got != "/usr/bin" {
		t.Errorf("unguarded PATH = %q, want it readable", got)
	}
	if !slices.Contains(e.childEnvironment(), "PATH=/usr/bin") {
		t.Error("PATH was stripped from the child's environment")
	}
	if got := e.get("SHOJIKU_BIN"); got != "" {
		t.Errorf("a SHOJIKU_ variable was still read: %q", got)
	}
}

func TestLookupsOnPassTheParentEnvironmentThroughUnchanged(t *testing.T) {
	source := []string{"SHOJIKU_FONT_DIR=/fonts", "PATH=/usr/bin"}
	e := &env{enabled: true, source: source}

	if got := e.childEnvironment(); !slices.Equal(got, source) {
		t.Errorf("childEnvironment() = %v, want the parent's %v", got, source)
	}
	if got := e.get("SHOJIKU_FONT_DIR"); got != "/fonts" {
		t.Errorf("get = %q, want /fonts", got)
	}
}

func TestABlankVariableReadsAsUnset(t *testing.T) {
	e := &env{enabled: true, source: []string{"SHOJIKU_BIN="}}

	if got := e.get("SHOJIKU_BIN"); got != "" {
		t.Errorf("get = %q, want a blank variable to read as unset", got)
	}
	if got := e.paths("SHOJIKU_FONT_DIR"); got != nil {
		t.Errorf("paths = %v, want nil for an unset variable", got)
	}
}

func TestPackDirectoriesComeFromTheEnvironmentAndExplicitConfigurationBeatsThem(t *testing.T) {
	separator := string(os.PathListSeparator)
	source := []string{"SHOJIKU_FONT_DIR=/a" + separator + separator + "/b"}

	fromEnv := &settings{config: config{}, env: &env{enabled: true, source: source}}
	if got := fromEnv.fontDirs(); !slices.Equal(got, []string{"/a", "/b"}) {
		t.Errorf("fontDirs = %v, want the environment's two entries with the blank dropped", got)
	}

	configured := &settings{
		config: config{fontDirs: []string{"/configured"}},
		env:    &env{enabled: true, source: source},
	}
	if got := configured.fontDirs(); !slices.Equal(got, []string{"/configured"}) {
		t.Errorf("fontDirs = %v, want the explicit value to win", got)
	}

	// An explicitly EMPTY list is a configuration, not a gap: it must not
	// fall back to the environment.
	none := &settings{
		config: config{localeDirs: []string{}},
		env:    &env{enabled: true, source: []string{"SHOJIKU_LOCALE_DIR=/from-env"}},
	}
	if got := none.localeDirs(); len(got) != 0 {
		t.Errorf("localeDirs = %v, want an explicit empty list to stay empty", got)
	}
}

func TestTheEnvironmentSuppliesTheTemplateRootAndExplicitConfigurationBeatsIt(t *testing.T) {
	root := fixtureTemplates(t)
	t.Setenv("SHOJIKU_TEMPLATE_ROOT", root)
	t.Setenv("SHOJIKU_BIN", engineBinary(t))

	fromEnv, err := NewClient()
	if err != nil {
		t.Fatalf("building a client from the environment: %v", err)
	}
	if fromEnv.TemplateRoot() != root {
		t.Errorf("root = %q, want the environment's %q", fromEnv.TemplateRoot(), root)
	}

	explicit, err := NewClient(WithTemplates("/elsewhere"))
	if err != nil {
		t.Fatalf("building a client with an explicit root: %v", err)
	}
	if explicit.TemplateRoot() != "/elsewhere" {
		t.Errorf("root = %q, want the explicit value to win", explicit.TemplateRoot())
	}

	off, err := NewClient(WithEnv(false), WithBinary(engineBinary(t)))
	if err != nil {
		t.Fatalf("building a hermetic client: %v", err)
	}
	if off.TemplateRoot() != "" {
		t.Errorf("root = %q, want lookups off to disable the environment", off.TemplateRoot())
	}
}

func TestAHermeticClientStillRendersWithoutInheritedPackDirectories(t *testing.T) {
	// The end-to-end shape of the flag: a client with lookups off renders
	// from what it was configured with, in a process whose environment names
	// a font directory that does not exist. If the child inherited it the
	// render would fail.
	t.Setenv("SHOJIKU_FONT_DIR", filepath.Join(t.TempDir(), "absent"))

	result, err := newTestClient(t).Generate(context.Background(), "receipt",
		map[string]any{"customer": map[string]any{"name": "Hermetic"}})

	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if !result.Success() {
		t.Fatalf("the render failed: %v", result.Failure())
	}
}

File: sdk/php/tests/EnvTest.php
<?php

declare(strict_types=1);

namespace Shojiku\Tests;

use PHPUnit\Framework\TestCase;
use Shojiku\Env;

final class EnvTest extends TestCase
{
    use EngineFixtures;

    public function testOneFlagGovernsEveryLookup(): void
    {
        $source = [
            'SHOJIKU_TEMPLATE_ROOT' => '/root',
            'SHOJIKU_FONT_DIR' => '/fonts',
            'SHOJIKU_BIN' => '/bin/shojiku',
        ];

        $on = new Env(enabled: true, source: $source);
        self::assertSame('/root', $on->get('SHOJIKU_TEMPLATE_ROOT'));
        self::assertSame(['/fonts'], $on->paths('SHOJIKU_FONT_DIR'));
        self::assertSame('/bin/shojiku', $on->get('SHOJIKU_BIN'));

        $off = new Env(enabled: false, source: $source);
        self::assertNull($off->get('SHOJIKU_TEMPLATE_ROOT'));
        self::assertSame([], $off->paths('SHOJIKU_FONT_DIR'));
        self::assertNull($off->get('SHOJIKU_BIN'));
    }

    public function testABlankVariableReadsAsUnset(): void
    {
        $env = new Env(enabled: true, source: ['SHOJIKU_LANG' => '']);

        self::assertNull($env->get('SHOJIKU_LANG'));
        self::assertNull($env->get('SHOJIKU_MISSING'));
    }

    public function testSeveralPathsInOneVariableSplitOnThePathSeparator(): void
    {
        $env = new Env(enabled: true, source: [
            'SHOJIKU_LOCALE_DIR' => '/a'.PATH_SEPARATOR.PATH_SEPARATOR.'/b',
        ]);

        self::assertSame(['/a', '/b'], $env->paths('SHOJIKU_LOCALE_DIR'));
    }

    public function testTheProcessEnvironmentIsTheDefaultSource(): void
    {
        putenv('SHOJIKU_TEMPLATE_ROOT=/from-process');

        try {
            self::assertSame('/from-process', (new Env(enabled: true))->get('SHOJIKU_TEMPLATE_ROOT'));
        } finally {
            putenv('SHOJIKU_TEMPLATE_ROOT');
        }
    }

    public function testTheChildInheritsTheEnvironmentWhenLookupsAreOn(): void
    {
        $env = new Env(enabled: true, source: ['PATH' => '/usr/bin', 'SHOJIKU_FONT_DIR' => '/fonts']);

        self::assertSame(['PATH' => '/usr/bin', 'SHOJIKU_FONT_DIR' => '/fonts'], $env->childEnvironment());
    }

    public function testTheChildLosesEverySHOJIKUVariableWhenLookupsAreOff(): void
    {
        // The clause a subprocess SDK owes that the in-process ones do not:
        // the engine is a CHILD PROCESS that reads SHOJIKU_FONT_DIR and
        // SHOJIKU_LOCALE_DIR itself, so a client that stopped reading them and
        // still let the child inherit them would only have moved the lookup
        // one process away.
        $env = new Env(enabled: false, source: [
            'PATH' => '/usr/bin',
            'SHOJIKU_FONT_DIR' => '/fonts',
            'SHOJIKU_TEMPLATE_ROOT' => '/root',
        ]);

        self::assertSame(['PATH' => '/usr/bin'], $env->childEnvironment());
    }

    public function testAHermeticClientDoesNotLetTheEngineReadTheEnvironmentEither(): void
    {
        // The same claim, end to end: a font directory the engine would
        // otherwise find in its own environment does not reach it. The
        // fixture template names `noto-sans`, so a render with no font
        // directory at all fails — which is what proves the variable did not
        // cross.
        putenv('SHOJIKU_FONT_DIR='.$this->fontDirs()[0]);

        try {
            $hermetic = $this->client(['fontDirs' => [], 'localeDirs' => [], 'env' => false]);
            self::assertTrue($hermetic->generate('receipt', [])->failed());

            $inheriting = $this->client(['fontDirs' => [], 'env' => true]);
            self::assertTrue($inheriting->generate('receipt', [])->success());
        } finally {
            putenv('SHOJIKU_FONT_DIR');
        }
    }
}
